### PR TITLE
feat(collector): upgrade the SGLang collector to 0.5.14

### PR DESCRIPTION
**Pull request by tianhaox**
## Summary

- supersede #1258 by replaying its SGLang 0.5.14 upgrade on top of #1302
- align stock SGLang attention, MLA/DSA, DSV4, GLM-5 sparse, GDN, mHC, GEMM/BMM, and MoE collectors with the pinned 0.5.14 runtime
- fail closed on persistence, CUDA Graph, quant-method, and grouped-inner errors instead of silently skipping, falling back, or writing mislabeled rows
- filter impossible DSV4 context hidden-state shapes at getter time and bind the exact retained inner manifest to scheduler/checkpoint identity
- record executed kernel provenance while keeping it producer-only; no SDK/Rust query changes
- track the fixed-version collector workflow, population rules, and Hopper/Blackwell handoff ledger

## Dependency and history

This PR depends on #1302. Its head commit `4c6cfebd` is an ancestor of this branch, so the diff against `main` includes #1302 until that PR merges.

The original #1258 commit `9a5fe012` was rebased as `f9c1e29c`, with conflict resolution following #1302's removal of the old exception-rule mechanism. The remaining commits are the subsequent SGLang 0.5.14 bring-up, backend/source audit, failure-accounting fixes, tests, and durable documentation.

## Scope

- exact stock framework contract: `sglang==0.5.14`
- image: `lmsysorg/sglang:v0.5.14-cu130@sha256:5027e95bf6ec536856b1b52a91d1f35ff5c564ab83e8a94758a169ff09bb8df3`
- no additional changes after #1302 under `collector/vllm`, `collector/trtllm`, `src`, `rust`, packaged data, or `.claude/rules`
- WideEP/special runtimes and other framework upgrades remain separate
- SM90 is the execution gate; SM100 is retained for the dependent B200 continuation; SM103/SM120 remain source-derived and hardware-unvalidated

## Important correctness changes

- DSA and DSV4 generation enter SGLang's production capture mode rather than measuring a sequential branch under a generic CUDA Graph
- standalone MLA no longer applies a legacy FlashMLA bound to FA3 cases
- sparse and grouped collectors preserve every inner failure and forbid unrecorded graph-to-eager fallback
- grouped DSV4 context tasks carry their getter-retained inner manifest, so task IDs and resume checkpoints identify the exact admitted work
- H20 DSV4 context uses the operation-specific BF16 lower bound `batch_size * sequence_length * hidden_size * 2 <= 80% * device_total_memory`; the model config supplies `hidden_size`, and no H20/SM/model shape constant is embedded
- GEMM, MLA-BMM, mHC, DSA, and framework-layer MoE rows derive provenance from the invoked path
- DSV4 FP4 experts on SM90 fail before timing when SGLang constructs an FP8 method; FP8 work is never persisted as MXFP4
- Marlin remains restricted to INT4-WO; NVFP4 and MXFP4 are distinct contracts

## Validation

- `390 passed`: collector unit/contract suite excluding the fresh-process parallel file
- `26 passed`: parallel worker/restart suite in a fresh process
- focused exact-image checks: `75 passed`
- Ruff check and format check pass
- `git diff --check` passes
- live H20-3e getter probe in the pinned image reproduced 88 outer tasks and 52,432 retained inner cells per CSA/HCA context op; batch 1024 retains through sequence 12,288
- branch head `e8ba45b1` is pushed to `origin/codex/sglang-0514-collector-only`

No post-#1302 GPU data is accepted yet. A benchmark smoke started after a clean preflight but overlapped with an unrelated vLLM job that subsequently occupied all GPUs; the task-owned SGLang container was stopped and its partial row rejected. Per release-owner direction, the code commit proceeded without waiting for another smoke. An uncontaminated smoke in a fresh namespace remains required before the new SM90 full collection begins.

## Summary by CodeRabbit

* **New Features**
  * Collector v2: improved planning with `--plan-only` and repeatable `--case-filter` for SM-aware op selection.
  * Added generation-time hardware capability floors and a hang/kill denylist to drop unsupported cases earlier.
  * SGLang collections are updated to use the pinned runtime behavior (0.5.14) with expanded attention/MoE routing metadata.
* **Bug Fixes**
  * Stricter fail-closed perf persistence and resume validation (checkpoint now binds framework version + SM; unresolved failures surface on resume).
* **Documentation**
  * Added collector upgrade playbook and expanded collector rule documentation for v2 behavior.
* **Tests**
  * Updated/added unit coverage for capability filtering, fail-closed behavior, and new plan/case selection contracts.

### CODE REVIEW

File: .claude/rules/collector/layer_permissions.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+then act. Violating these boundaries is how the retired sm_exceptions rule
+engine grew to 1400 lines of shape-sniping rules.
+
+## The one-line architecture
+
+```text
+plan cases = dedup( expand(base_ops sweep grids) ∪ expand(model_case_values shapes) )
+runnable   = plan cases ∩ hardware capability floors  −  hang denylist
+result     = perf rows  +  classified failure records   (failure is DATA, not a defect)
+```
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply .claude/rules/collector/layer_permissions.md:22]: This file is #1302's content (human-owned rule policy per its own module-boundary section), so the equation change belongs to that PR rather than this collector-only continuation. Noted for the #1302 review: the runnable set is `(base ∪ model shapes) ∩ capability floors − denylist − registry maturity markers`. Resolving here to keep this PR's review scoped.

File: collector/case_generator.py [hunk 2]
```diff
@@ -155,41 +164,183 @@ def _head_profiles(
...
+                    86: "triton",
+                    89: "triton",
+                    90: "fa3",
+                    100: "trtllm_mha",
+                    103: "trtllm_mha",
+                    120: "flashinfer",
+                }.get(sm_version)
+            if kernel_source is None:
+                raise ValueError(f"No SGLang 0.5.14 attention backend mapping for SM{sm_version}")
+            kernel_source = str(kernel_source)
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/case_generator.py:354]: Addressed earlier in f8b299b1 (default map now mirrors `server_args._get_default_attn_backend`), and 6f85f6f8 adds the exact file:line citations (server_args.py:4407-4455 at image source 49e384ce), including the sink-aware flashinfer/triton split for the non-Hopper/non-SM100 branch.

File: collector/collect.py [hunk 3]
```diff
@@ -503,18 +481,18 @@ def worker(
...
         if current_task_ids is not None:
             current_task_ids[device_id] = task_id
 
         try:
             worker_logger.debug(f"Starting task {task_id}")
-            func(*task, device=device)
+            result = func(*task, device=device)
+            if isinstance(result, int) and result == EXIT_CODE_RESTART:
+                raise SystemExit(EXIT_CODE_RESTART)
             worker_logger.debug(f"Completed task {task_id}")
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/collect.py:503]: Fixed in 8c8c2290: the executor no longer interprets any int result as a restart. GDN/MoE return a dedicated `WORKER_RESTART` signal object (helper.WorkerRestartSignal); `sys.exit(EXIT_CODE_RESTART)` is unchanged; a new parallel-run test locks that a plain int 10 result (e.g. a logged-row count, as collect_mla_module returns) completes normally without recycling the worker.

File: collector/sglang/collect_attn.py [hunk 4]
```diff
@@ -325,18 +386,49 @@ def run_attention_torch(
...
+            80: "triton",
+            86: "triton",
+            89: "triton",
+            90: "fa3",
+            100: "trtllm_mha",
+            103: "trtllm_mha",
+            120: "flashinfer",
+        }.get(sm_version)
+        if attn_backend_name is None:
+            raise ValueError(f"No SGLang 0.5.14 attention backend mapping for SM{sm_version}")
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/collect_attn.py:427]: Fixed in 6f85f6f8: the map now cites `python/sglang/srt/server_args.py:4407-4455` (image source 49e384ce) per branch, and documents that per-model deviations (Qwen3.5 hybrid-GDN→triton, server_args.py:4188-4211; NemotronH→flashinfer, arg_groups/nemotron_h_hook.py:60-62) arrive only through the profile's `sglang_backends` map, never through this default table. B200 execution probes ran every declared SM100 route (trtllm_mha/triton/flashinfer) with persisted rows.

File: collector/sglang/collect_gemm.py [hunk 5]
```diff
@@ -78,6 +77,10 @@ def get_gemm_test_cases():
         for gemm_type in gemm_list:
             if (gemm_type == "nvfp4" or gemm_type == "fp8_block") and (n < 128 or k < 128):
                 continue
+            if gemm_type == "nvfp4" and (n % 128 != 0 or k % 64 != 0):
+                # The 0.5.14 FlashInfer path uses the modelopt 128x4
+                # block-scale layout without synthetic padding.
+                continue
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/collect_gemm.py]: Split in 6f85f6f8. (a) The NVFP4 alignment `continue` was dead code (0/483 unique (n,k) in the shared sweep are misaligned) and is now a classified runtime raise in run_gemm citing the modelopt 128x4 block-scale layout — no plan delta. (b) The inherited n<128/k<128 filter is parked as a dated FIXME(kernel-limit): removing it adds 6,216 specs per affected mode on BOTH platforms (SM90 fp8_block included), i.e. a shared coverage-contract change that must not ride along this Blackwell continuation; it is queued for the next explicit coverage review / version-bump audit.

File: collector/sglang/collect_gemm.py [hunk 6]
```diff
@@ -144,16 +147,21 @@ def run_gemm(gemm_type, batch_size, N, K, *, perf_filename, device="cuda:0"):  #
...
+    fp4_backend = None
+    if gemm_type == "nvfp4":
+        if not HAS_FLASHINFER_FP4:
+            raise RuntimeError("SGLang NVFP4 GEMM requires FlashInfer FP4 quantization support")
+        if sm_version in {100, 103}:
+            fp4_backend = "cute-dsl"
+        elif sm_version == 120:
+            fp4_backend = "cutlass"
+        else:
+            raise ValueError(f"SGLang NVFP4 dense GEMM is not implemented for SM{sm_version}")
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/collect_gemm.py:181]: Fixed in 6f85f6f8: the pin now cites SGLang 0.5.14 `initialize_fp4_gemm_config` (python/sglang/srt/layers/quantization/fp4_utils.py:148-161 at image source 49e384ce): "auto" resolves to flashinfer_cutedsl (mm_fp4 "cute-dsl") when is_sm100_supported() (major 10 → SM100/103), marlin on SM80-89, and flashinfer_cutlass (mm_fp4 "cutlass") otherwise, which is the SM120 path — matching the collector's table exactly.

File: collector/sglang/deepseekv4_sparse_modules.py [hunk 7]
```diff
@@ -276,23 +275,25 @@ def _device_num_sms(device: str | torch.device) -> int:
...
         return importlib.util.find_spec("flash_mla") is not None
     if kernel == "paged_mqa_logits":
-        if importlib.util.find_spec("deep_gemm") is None:
-            return False
-        if torch.cuda.is_available():
-            major, _minor = torch.cuda.get_device_capability()
-            if major >= 12:
-                return False
-        return True
+        return importlib.util.find_spec("deep_gemm") is not None
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/deepseekv4_sparse_modules.py]: Fixed in bdd80363: both `_dsv4_sparse_kernel_supported` and `_dsv4_topk_kernel_supported` are deleted. Pre-Hopper moves to `op_min_sm: 90` capability floors (same SM90+ kernel-library rationale and precedent as the dsa_*_module floors); SM120 is parked by registry `unverified_sms=(120,)`; and the flash_mla-less HCA/CSA kernel collectors — no platform has ever produced rows for them in the pinned image — are parked whole-op `unverified=True` so the skip is an explicit summary entry instead of a silent empty plan. Only plan delta: those two ops go from silent 0 to explicit unverified skip; all other counts unchanged on SM90/100 (re-enumerated on B200).

File: collector/sglang/deepseekv4_sparse_modules.py [hunk 8]
```diff
@@ -1022,28 +1028,51 @@ def run_dsv4_sparse_kernel_worker(
...
+    else:
+        _seen = set(ctx_shapes)
+        shapes = [(*shape, None) for shape in ctx_shapes]
+        shapes.extend((*shape, None) for shape in dec_shapes if shape not in _seen)
     # this task owns one bs (collect.py distributes bs across GPU workers)
-    shapes = [(prefix, isl, bs) for (prefix, isl, bs) in shapes if bs == bs_only]
+    shapes = [(prefix, isl, bs, variant) for (prefix, isl, bs, variant) in shapes if bs == bs_only]
     if not shapes:
         print(f"[dsv4-sparse {kernel} bs={bs_only}] no shapes; skipping.")
         return
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/deepseekv4_sparse_modules.py]: Fixed in bdd80363: a queued (kernel, bs) task that resolves no shapes now raises `RuntimeError("dsv4-sparse {kernel} bs={bs}: queued task resolved no shapes")` so the case-plan inconsistency is recorded as a classified failure instead of a clean completion.

File: collector/sglang/glm5_dsa_sparse_modules.py [hunk 9]
```diff
@@ -550,18 +690,29 @@ def run_glm5_dsa_sparse_kernel_worker(
     if not shapes:
         print(f"[{label}-sparse {kernel} bs={bs_only}] no shapes; skipping.")
         return
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/glm5_dsa_sparse_modules.py]: Fixed in bdd80363: same fail-closed treatment as the DSV4 sparse worker — an empty resolved shape list raises a classified RuntimeError instead of returning silently.

File: collector/sglang/glm5_dsa_sparse_modules.py [hunk 10]
```diff
@@ -579,10 +730,18 @@ def run_glm5_dsa_sparse_kernel_worker(
...
+    summary = f"ok={n_ok} error={error_count} skip=0 total={len(shapes)}"
+    print(f"  {kernel}: {summary}")
+    if not n_ok or failures:
+        details = "\n- ".join(failures) if failures else "no inner shapes produced a row"
+        raise RuntimeError(f"{label}-sparse {kernel} bs={bs_only}: {summary}; failures:\n- {details}")
 
 
 def _glm5_sparse_kernel_cases(kernel):
+    if get_sm_version() not in {90, 100, 103}:
+        return []
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/glm5_dsa_sparse_modules.py]: Fixed in bdd80363: the `get_sm_version() not in {90, 100, 103}` gate is deleted from `_glm5_sparse_kernel_cases`. Pre-Hopper is dropped by the new `op_min_sm: 90` floors for the three glm5 ops (DeepGEMM fp8_mqa_logits / FlashMLA sparse family), and SM120's different indexer API is represented by registry `unverified_sms=(120,)` markers. A test locks that the getter itself carries no SM predicate and that floors+markers exist. (The remaining SM branch in `_selected_glm5_models` picks WHICH artifact represents the family — the NVFP4 checkpoint is only constructible on SM100+ — which is dispatch, not case removal.)

File: collector/sglang/collect_dsv4_attn.py [hunk 11]
```diff
@@ -1374,52 +1745,47 @@ def _subprocess_entry(
...
+        for cur_prefix in prefix_values:
+            pairs = [
+                (bs, sl)
+                for bs, sl in _filter_pairs(mode, [batch_size], sl_grid)
+                if _is_valid_shape(mode, bs, sl, cur_prefix)
+                and (max_position_embeddings is None or cur_prefix + sl <= max_position_embeddings)
+            ]
+            if not pairs:
+                print(f"[dsv4-flash] no valid sl values for mode={mode}, bs={batch_size}, prefix_len={cur_prefix}")
+                continue
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply collector/sglang/collect_dsv4_attn.py]: Fixed in bdd80363: the per-prefix `continue` is now a classified RuntimeError, so the worker-side grid path (generation / manual CLI; context uses the getter-retained manifest) fails on the FIRST unresolvable prefix group instead of succeeding with partial coverage. The existing contract test was updated to the stricter message.

**Comment by tianhaox** [comment]
@coderabbitai review

File: tests/unit/collector/test_sglang_glm5_sparse_0514.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+    _, op_min_sm = _load_capabilities()
+    floors = ("dsv4_paged_mqa_logits_module", "dsv4_hca_attn_module", "dsv4_csa_attn_module", "dsv4_csa_topk_calib")
+    for op in floors:
+        assert op_min_sm.get(op) == 90
+    by_op = {entry.op: entry for entry in REGISTRY}
+    assert by_op["dsv4_paged_mqa_logits_module"].unverified_sms == (120,)
+    assert by_op["dsv4_csa_topk_calib"].unverified_sms == (120,)
+    assert by_op["dsv4_hca_attn_module"].unverified is True
+    assert by_op["dsv4_csa_attn_module"].unverified is True
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply tests/unit/collector/test_sglang_glm5_sparse_0514.py:141]: Fixed in e849c8c5: added module-level `pytestmark = pytest.mark.unit`; the file now runs under `-m unit` (10 passed).

**Comment by tianhaox** [comment]
Superseded by #1342 — same branch content rebased onto latest main, now carrying the collected h200_sxm/b200_sxm 0.5.14 performance data on an upstream branch (`feat/sglang-0514-collector-and-data`) per the data-PR requirement.
